specs/experimental/contracts/cannon/mips64.md

```diff
@@ -0,0 +1,194 @@
+  - [stateVersion](#stateversion)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+The MIPS64 contract is an on-chain implementation of a MIPS64 virtual machine that executes a single instruction step.
+It serves as the state transition verification component in the Cannon fault proof system, enabling on-chain validation
+of off-chain computation by proving correct execution of MIPS64 instructions within dispute games.
+
+## Definitions
+
+### Thread Stack
+
+A hash onion structure representing a stack of thread states. Each thread is committed by computing
+`Keccak256(previous_root ++ thread_hash)`, where the empty stack is represented by
+`Keccak256(bytes32(0) ++ bytes32(0))`. This construction allows succinct commitment to all threads using a single
+bytes32 value. See the [Cannon Fault Proof VM specification](../../fault-proof/cannon-fault-proof-vm.md#thread-stack-hashing)
+for detailed hashing mechanics.
+
+### Memory Reservation
+
+A mechanism enabling atomic read-modify-write operations through Load Linked and Store Conditional instructions.
+A reservation tracks a memory address, the owning thread, and the operation size (32-bit or 64-bit). Any memory write
+to the reserved address clears the reservation, ensuring atomicity across concurrent threads.
+
+### Context Switch
+
+The process of preempting the currently active thread and activating another thread from the thread stacks.
+Context switches occur voluntarily through specific syscalls or forcibly after executing a quantum of steps without
+yielding.
+
+### State Version
+
+An identifier specifying the state transition rules implemented by the contract. The MIPS64 contract supports
+versions 7 (VersionMultiThreaded64_v4) and 8 (VersionMultiThreaded64_v5), which may differ in syscall behavior or
+state field interpretations.
+
+## Assumptions
+
+### a01-001: PreimageOracle provides valid data
+
+The PreimageOracle contract returns valid preimage data when queried during read syscalls. If the oracle returns
+incorrect or malicious data, the VM state transition will be incorrect, potentially causing invalid state roots to be
+accepted in dispute games.
+
+#### Mitigations
+
+- The PreimageOracle contract is trusted and governance-controlled
+- Preimage data is deterministically derived from the preimage key
+- Dispute game participants can independently verify preimage correctness off-chain
+
+### a01-002: Proof data is well-formed
+
+Callers provide structurally valid proof data including memory proofs and thread witness data. While the contract
+validates proof correctness against state commitments, it assumes the proof structure matches expected formats and
+offsets.
+
+#### Mitigations
+
+- Proof validation logic verifies merkle proofs against state roots
+- Invalid proofs cause reverts rather than incorrect state transitions
+- Off-chain proof generation is deterministic and can be independently verified
+
+## Invariants
```

Comment:
devinai Only invariants that matter is that it's deterministic and that it correctly, exactly implements the MIPS64 spec (MIPS64 spec divergences are HIGH severity unless those spec divergences are not actually relied on by the Go compiler, which does happen sometimes, as in the case of certain unimplemented opcodes, in which case it is a Low severity bug).